### ISSUE TITLE
Tracking Issue: Make Harnesses Optional

### ISSUE THREAD

**Issue by carolynzech** [model-checking/kani#3832]
Requested feature: Make proof harnesses optional, when possible
Use case: There are plenty of "boilerplate harnesses" that look something like this:

```rust
fn foo(x: u8, y: u8) { ... }

#[kani::proof]
fn foo_proof() {
  let x = kani::any();
  let y = kani::any();
  foo(x, y);
}
```
There's nothing surprising in this harness--the user just creates arbitrary variables for each argument of the function they're verifying, and then calls the function. Ideally, the user shouldn't have to write this harness at all; Kani should detect that each of the arguments implements `Arbitrary`, then attempt to verify it automatically.

## Tasks

(We define "standard harness" as a `#[kani::proof]` harness and a "contract harness" as a `#[kani::proof_for_contract]` harness).
I'll break these tasks down into subtasks as I start working, but a road map for now:

- [ ] Leverage the work from https://github.com/model-checking/kani-auto-harness to autogenerate standard harnesses for types that _trivially_ implement `Arbitrary`. (Write the harnesses to a temporary file and run them automatically).
- [ ] Detect which functions have contracts and autogenerate contract harnesses for them.
- [ ] Extend the harness autogeneration to support more complex use cases (c.f [known issues](https://github.com/model-checking/kani-auto-harness/issues))--namely, pointer and loop support.
- [ ] Transition away from writing harnesses to a file, then running them: we shouldn't have to write literal harness Rust code at all. We should just decide which functions we want to verify and verify them, without needing to find some function with a `#[kani::proof]` annotation as an entrypoint for verification.

## Open Questions
  - UI: Support Miri-esque error messages where we say which function caused the problem, rather than which harness, ideally pointing to the problematic line(s) of code.
  - Performance: If we try to verify everything that we can verify, it may take quite a long time. Should we try to be more discerning about what we verify, e.g., introducing some kind of ordering (e.g., look for unsafe functions first, then safe abstractions?) Should this feature have an automatic harness timeout, or perhaps an attribute you can put on a function to opt out of automatic verification?
  - Reasonable defaults: For types that do not trivially implement `Arbitrary` (e.g., pointers), we need to decide:
    - Do reasonable default values exist?
    - If so, what are they, and how do we communicate our default assumptions clearly so that users don't have a false sense of security if their code verifies?

     Similarly, for code with loops, we can statically detect the presence of the loop, but most of the time we won't be able to statically determine an unwinding bound. So we need to decide if it's worth it to pick an arbitrary number, or just leave loops out of this for now.

**Comment by carolynzech** [comment]
Note from #3874 -- Should log metadata about which functions were skipped, then print those functions in the driver, ideally with a note about why they were skipped.

**Comment by carolynzech** [comment]
For now, we don't support automatic harnesses for closures for two reasons:

1. We can't tell if a closure comes from Kani instrumentation (#3809).
2. Generating an `Instance` from a closure `CrateItem` isn't working, i.e., the `Instance::try_from(fn_item)` call in #3809 fails on a closure like:
```rust
const MY_CLOSURE: fn(u8) -> u8 = |x: u8| x + 1;
```
whereas
```rust
const MY_CLOSURE: fn(u8) -> u8 = |x: u8| x + 1;

#[kani::proof]
fn harness() {
  MY_CLOSURE(kani::any());
}
```
runs fine--Kani finds the `Instance` for the closure just fine during its reachability analysis. I don't think this is necessarily worthy of a separate issue yet because it may just be something simple that I'm doing wrong, and it's a moot point as long as #3809 is still open.

**Comment by carolynzech** [comment]
Apparently for tuple structs, Rust will insert a constructor function in MIR, e.g.
```rust
struct TupleStruct(u32);
```
generates:
```
fn TupleStruct(_1: u32) -> TupleStruct {
    let mut _0: TupleStruct;

    bb0: {
        _0 = TupleStruct(move _1);
        return;
    }
}
```
(see [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=657699b34719727d4f58878d4d12c814)).

This means that in autoharness, we find and verify this function:
```
Autoharness Summary:
+-------------------+---------------------------+---------------------+
| Selected Function | Kind of Automatic Harness | Verification Result |
+=====================================================================+
| TupleStruct       | #[kani::proof]            | Success             |
+-------------------+---------------------------+---------------------+
Complete - 1 successfully verified functions, 0 failures, 1 total.
```

More confusingly, with #3942 merged, if we have:
```rust
struct TupleStruct(Vec<u32>);
```
we get:
```
+------------------+-----------------------------------------------------+
| Skipped Function | Reason for Skipping                                 |
+========================================================================+
| TupleStruct      | Missing Arbitrary implementation for argument(s): _ |
+------------------+-----------------------------------------------------+
```

The correct behavior would be to skip over these constructors entirely, since they are inserted by the Rust compiler. I don't see a great way of doing that:
- We could do something hacky like "if the name of the function is the same as a struct in the same namespace, skip it," but I worry about a bug in that logic causing legitimate functions to be ignored
- This may also be a point in favor of @zhassan-aws's suggestion in https://github.com/model-checking/kani/pull/3874#discussion_r1945301042, since this constructor is a private function.

**Comment by carolynzech** [comment]
Consider adding a flag e.g. `--write-to-file` that should write Rust harnesses back to the file.

**Comment by carolynzech** [comment]
Now that we log which functions we skipped, it's common for us to say that we skipped the `Clone`, `Hash`, etc. implementations because they take a reference, which we don't support. Once we support references, this will be an issue--we should have some principled way of skipping over standard library traits like this.

**Comment by carolynzech** [comment]
Consider adding a flag to only generate harnesses for unsafe functions or safe abstractions.

**Comment by carolynzech** [comment]
Kani will also add autoharnesses for nested functions if the outer function has a contract, c.f. https://github.com/model-checking/verify-rust-std/pull/350#discussion_r2091698600
